domain/user/state/state.go

```diff
@@ -0,0 +1,335 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"time"
+
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/user"
+	"github.com/juju/juju/domain"
+	domainuser "github.com/juju/juju/domain/user"
+	usererrors "github.com/juju/juju/domain/user/errors"
+	databaseutils "github.com/juju/juju/internal/database"
+)
+
+// State represents a type for interacting with the underlying state.
+type State struct {
+	*domain.StateBase
+}
+
+// NewState returns a new State for interacting with the underlying state.
+func NewState(factory database.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+func addUser(ctx context.Context, tx *sql.Tx, uuid domainuser.UUID, user user.User, creatorUUID domainuser.UUID) error {
+
+	addUserQuery := "INSERT INTO user (uuid, name, display_name, created_by_uuid, created_at) VALUES (?, ?, ?, ?, ?)"
+
+	_, err := tx.ExecContext(ctx, addUserQuery,
+		uuid.String(),
+		user.Name,
+		user.DisplayName,
+		creatorUUID.String(),
+		time.Now(),
+	)
+	if databaseutils.IsErrConstraintUnique(err) {
+		return usererrors.AlreadyExists
+	} else if databaseutils.IsErrConstraintForeignKey(err) {
+		return usererrors.UserCreatorUuidNotFound
+	} else if err != nil {
+		return fmt.Errorf("cannot insert user %q: %w", user.Name, err)
+	}
+
+	return nil
+}
+
+func setPasswordHash(ctx context.Context, tx *sql.Tx, userUUID domainuser.UUID, passwordHash string, salt []byte) error {
+	defineUserAuthenticationQuery := `
+INSERT INTO user_authentication (user_uuid, disabled) 
+VALUES (?, ?) 
+ON CONFLICT(user_uuid) DO UPDATE SET user_uuid = excluded.user_uuid, disabled = excluded.disabled
```

Comment:
I think we have a logic problem here. We don't want todo `disabled = excluded.disabled`. The reason being that if the user_authentication record already exists and it is disabled we are about to update it false and make them active again. This isn't a documented side effect of the func and would introduce security problems for us.

Safest bet is to just drop the disabled update all together from the update. Actually this whole statement we can update to just.

```sql
SERT INTO user_authentication (user_uuid, disabled) 
VALUES (?, ?) 
ON CONFLICT(user_uuid) DO NOTHING
```